### PR TITLE
TIP-307: Fix issues with PHP 7 and MongoDB 2.6

### PR DESCRIPTION
**Pull request by nidup**
To fix the following issue with PHP7 and MongoDB 2.6 enforcing several limitations on indexes.

`̀``
[2016-07-26 10:03:20] batch.ERROR: Encountered an error executing the step: insertDocument :: caused by :: 17280 Btree::insert: key too large to index, failing pim_ma_pcdm.pim_catalog_product.$normalizedData.description-en_US-print_1 1490 { : "<b>Streamlined Reliability</b>\nA smart, reliable option for bringing duplex printing, copying, scanning and high-speed faxing into one machine, with ..." }  
An error occurred during the fixtures execution.
Error #17280 in class MongoCursorException: insertDocument :: caused by :: 17280 Btree::insert: key too large to index, failing pim_ma_pcdm.pim_catalog_product.$normalizedData.description-en_US-print_1 1490 { : "<b>Streamlined Reliability</b>\nA smart, reliable option for bringing duplex printing, copying, scanning and high-speed faxing into one machine, with ..." }                                                                                                                                                          
#0 /home/jjanvier/workspaces/phpstorm/akeneo/pim_master/pcdm/vendor/alcaeus/mongo-php-adapter/lib/Mongo/MongoCollection.php(350): Alcaeus\MongoDbAdapter\ExceptionConverter::toLegacy(Object(MongoDB\Driver\Exception\BulkWriteException), 'MongoResultExce...')                                                                                            
#1 /home/jjanvier/workspaces/phpstorm/akeneo/pim_master/pcdm/vendor/doctrine/mongodb/lib/Doctrine/MongoDB/Collection.php(1022): MongoCollection->batchInsert(Array, Array)    
#2 /home/jjanvier/workspaces/phpstorm/akeneo/pim_master/pcdm/vendor/doctrine/mongodb/lib/Doctrine/MongoDB/Collection.php(158): Doctrine\MongoDB\Collection->doBatchInsert(Array, Array)                                                                                                                                                                     
#3 
`̀``

Need to merge this first https://github.com/akeneo/pim-community-dev/pull/4741

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Y
| Added Behats                      | N
| Changelog updated                 | Y
| Review and 2 GTM                  | Y
| Micro Demo to the PO (Story only) | N
| Migration script                  | Y
| Tech Doc                          | N

### CODE REVIEW

File: UPGRADE-1.6.md [hunk 1]
```diff
@@ -326,6 +326,20 @@ We provide doctrine migration to handle this change in your existing data.
...
 
+##  PHP7
+
+We continued our effort regarding Akeneo PIM PHP7 support.
+We're happy to announce that PHP7 is now usable in experimental mode for both Cli and Web.
+Experimental means that we manage to install and use the PIM but due to missing tests in our functional matrix we can't engage to support it.
+
+MongoDB 2.6 enforce several limitations, for instance, the Index Key Limit, https://docs.mongodb.com/v2.6/reference/limits/#Index-Key-Limit.
+
+This limitation impact indexes, when you use PHP7, please dump your database and run the command app/console 'pim:product:ensure-mongodb-indexes'.
```
Comment (jjanvier) [inline UPGRADE-1.6.md]: impactS

File: src/Pim/Bundle/CatalogBundle/Command/MongoDBIndexCreatorCommand.php [hunk 2]
```diff
@@ -43,6 +44,9 @@ protected function execute(InputInterface $input, OutputInterface $output)
             return -1;
         }
 
+        $indexPurger = $this->getIndexPurger();
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Command/MongoDBIndexCreatorCommand.php]: is it normal to do this here?
shouldn't it be a different command?

File: UPGRADE-1.6.md [hunk 3]
```diff
@@ -326,6 +326,20 @@ We provide doctrine migration to handle this change in your existing data.
 
 TODO: Add links to the docs.akeneo.com regarding this format update.
 
+##  PHP7
+
+We continued our effort regarding Akeneo PIM PHP7 support.
+We're happy to announce that PHP7 is now usable in experimental mode for both Cli and Web.
+Experimental means that we manage to install and use the PIM but due to missing tests in our functional matrix we can't engage to support it.
```
Comment (phaseinducer) [inline UPGRADE-1.6.md]: engage to -> commit to

File: UPGRADE-1.6.md [hunk 4]
```diff
@@ -326,6 +326,20 @@ We provide doctrine migration to handle this change in your existing data.
 
 TODO: Add links to the docs.akeneo.com regarding this format update.
 
+##  PHP7
+
+We continued our effort regarding Akeneo PIM PHP7 support.
+We're happy to announce that PHP7 is now usable in experimental mode for both Cli and Web.
```
Comment (phaseinducer) [inline UPGRADE-1.6.md]: CLI?

**Comment by jjanvier** [comment]
gtm once fixed

File: UPGRADE-1.6.md [hunk 5]
```diff
@@ -326,6 +326,20 @@ We provide doctrine migration to handle this change in your existing data.
...
+Experimental means that we manage to install and use the PIM but due to missing tests in our functional matrix we can't engage to support it.
+
+MongoDB 2.6 enforce several limitations, for instance, the Index Key Limit, https://docs.mongodb.com/v2.6/reference/limits/#Index-Key-Limit.
+
+This limitation impact indexes, when you use PHP7, please dump your database and run the command app/console 'pim:product:ensure-mongodb-indexes'.
+
+This command will remove existing indexes from product collection and will add new indexes with relevant types.
+
+Your attributes using `text` backend type will now use `hashed` index type supporting more than 1024 chars.
```
Comment (phaseinducer) [inline UPGRADE-1.6.md]: What was the behavior before? I don't fully understand this sentence. :/

Comment (nidup) [inline UPGRADE-1.6.md]: it only created ascendant indexes (aka 1) and it silently fails when the data to index is > 1024 chars in text attributes :smile: (i'll complete the desc)

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/IndexCreator.php [hunk 6]
```diff
@@ -24,6 +25,12 @@ class IndexCreator
     /** @staticvar int 64 is the MongoDB limit for indexes (not configurable) */
     const MONGODB_INDEXES_LIMIT = 64;
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/IndexCreator.php]: Please remove it, I can't stop having warnings about this during install.

File: src/Pim/Bundle/CatalogBundle/Command/MongoDBIndexCreatorCommand.php (hunk 2 above)
Comment (nidup) [inline src/Pim/Bundle/CatalogBundle/Command/MongoDBIndexCreatorCommand.php]: I hesitated but imvho, if we split, one couldn't / shouldn't be used without the other. The purpose of the command is to ensure that your indexes are well created, don't see the point of keeping previously created indexes, as project integrator, if you need to implement a different logic of priorities for index creation, you should override the IndexCreator.

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/IndexCreator.php (hunk 6 above)
Comment (nidup) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/IndexCreator.php]: without the check we made in ensureIndexes method an exception is raised and stop the install, we could add warning messages and not errors in the log but imho it seems weird to hide the fact that some indexes are missing.

Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/IndexCreator.php]: Sorry, it was just a troll ^^

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/IndexCreator.php [hunk 7]
```diff
@@ -24,6 +25,12 @@ class IndexCreator
...
 
+    /** @staticvar int the default MongoDB index type */
+    const ASCENDANT_INDEX_TYPE = 1;
+
+    /** @staticvar int the hash MongoDB index type */
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/IndexCreator.php]: string?

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/IndexCreator.php [hunk 8]
```diff
@@ -252,12 +265,17 @@ protected function addFieldsFromOption(array $fields)
 
     /**
      * Ensure indexes on the provided field names.
-     * Indexed are created in background and the PHP process does not
-     * wait for the completion of the index creation (w at 0)
      *
-     * @param array $fields
+     * Indexed are created in background and the PHP process does not wait for the completion of the index creation
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/IndexCreator.php]: IndexeS

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/IndexCreator.php (hunk 6 above)
Comment (nidup) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/IndexCreator.php]: a good one, the fact it was written makes me hesitated :wink:  :trollface: #meur

**Comment by phaseinducer** [comment]
GTM after fix

**Pull request merged**